### ISSUE TITLE
`unimplemented_default_member`

### ISSUE THREAD

**Issue by mladedav** [rust-lang/rust-clippy#12727]
### What it does

Forces useers to implement all methods in a trait, including all methods with default implementations.

This would be allow-by-default restriction lint. This makes sense to turn on only in specific scenarios.

### Advantage

- Users will not be surprised by an addition of new default members in traits if they wish to always implement all of them.

### Drawbacks

- There is no way to turn it on or off on a method-by-method basis so the user would sometimes be forced to copy the default implementation if that's the implementation they want to use. They would lose on any potential updates to its implementation.

### Example

```rust
trait Record {
  fn record_debug(field: &dyn Debug);

  fn record_u64(field: u64) {
    record_debug(field);
  }

  fn record_i64(field: i64) {
    record_debug(field);
  }
}
```

```rust
trait HasRecord {
    type MyRecord: Record;
    fn my_record() -> Self::MyRecord;
}

impl Record for T where T: HasRecord {
  fn record_debug(field: &dyn Debug) {
     Self::my_record().record_debug();
  }

  fn record_u64(field: u64) {
     Self::my_record().record_u64()
  }

  fn record_i64(field: i64) {
     Self::my_record().record_i64()
  }
}
```

Now if the `Record` trait is extended with a new method with default implementation such as `record_str(str: &str)` the default implementation would be bad for implementations leveraging the `HasRecord` type. As I user, I'd like a way to tell clippy that I want to be notified, if any method is not explicitly defined in a given `impl` block.

**Comment by Alexendoo** [comment]
Good news, we have a lint for this: https://rust-lang.github.io/rust-clippy/master/index.html#/missing_trait_methods

**Comment by mladedav** [comment]
Thanks and sorry for the noise. I did try to find it (unimplemented, trait and keywords like that) but didn't find it.
